lib/sqlitex/query.ex

```diff
@@ -0,0 +1,54 @@
+defmodule Sqlitex.Query do
+  def bind_values({:error, _}=error, _opts), do: error
+  def bind_values({:ok, statement}, opts) do
+    params = params_from_opts(opts)
+    case :esqlite3.bind(statement, params) do
+      {:error,_}=error -> error
+      :ok -> {:ok, statement}
+    end
+  end
+
+  def execute({:error,_}=error, _opts), do: {error, {}, {}, []}
+  def execute({:ok, statement}, opts) do
+    into = into_from_opts(opts)
+    types = :esqlite3.column_types(statement)
+    columns = :esqlite3.column_names(statement)
+    data = :esqlite3.fetchall(statement)
+    {types, columns, data, into}
+  end
+
+  def prepare(sql, database), do: :esqlite3.prepare(sql, database)
+
+  def to_rows({_,_,{:error,_}=error,_}), do: error
```

Comment:
You said you dislike the tuple here.  I don't think it makes too much of a difference, but could you prepend everything with the following match:

```elixir
def to_rows({types, cols, rows, into}), do: to_rows(types, cols, rows, into)
```

and then deal with four separate arguments in the remaining functions?